internal/core/src/clustering/KmeansClustering.h

```diff
@@ -0,0 +1,120 @@
+            storage::FileManagerContext());
+
+    void
+    Run() override;
+
+    ClusteringResult
+    GetClusteringResult() override {
+        return cluster_result_;
+    }
+
+    // for unit test
+    inline std::string
+    GetRemoteCentroidsObjectPrefix() const {
+        auto index_meta_ = file_manager_->GetIndexMeta();
+        auto field_meta_ = file_manager_->GetFieldDataMeta();
+        return file_manager_->GetChunkManager()->GetRootPath() + "/files" +
+               std::string(ANALYZE_ROOT_PATH) + "/" +
+               std::to_string(index_meta_.build_id) + "/" +
+               std::to_string(index_meta_.index_version) + "/" +
+               std::to_string(field_meta_.collection_id) + "/" +
+               std::to_string(field_meta_.partition_id) + "/" +
+               std::to_string(field_meta_.field_id);
+    }
+
+    inline std::string
+    GetRemoteCentroidIdMappingObjectPrefix(int64_t segment_id) const {
+        auto index_meta_ = file_manager_->GetIndexMeta();
+        auto field_meta_ = file_manager_->GetFieldDataMeta();
+        return file_manager_->GetChunkManager()->GetRootPath() + "/files" +
+               std::string(ANALYZE_ROOT_PATH) + "/" +
+               std::to_string(index_meta_.build_id) + "/" +
+               std::to_string(index_meta_.index_version) + "/" +
+               std::to_string(field_meta_.collection_id) + "/" +
+               std::to_string(field_meta_.partition_id) + "/" +
+               std::to_string(field_meta_.field_id) + "/" +
+               std::to_string(segment_id);
+    }
+
+ private:
+    void
+    StreamingAssignandUpload(const T* centroids);
+
+    // bool to indicate if the whole segment data are fetched
+    bool
+    FetchSegmentData(uint8_t* buf,
+                     int64_t expected_train_size,
+                     const std::vector<std::string>& files,
+                     const int64_t num_rows,
+                     int64_t& offset);
+
+    // given all possible segments, sample data to buffer
+    std::unique_ptr<uint8_t[]>
+    SampleTrainData(
+        const std::vector<int64_t>& segment_ids,
+        const std::map<int64_t, std::vector<std::string>>& segment_file_paths,
+        const std::map<int64_t, int64_t>& segment_num_rows,
+        int64_t expected_train_num);
+
+    // transform centroids result to PB format for future usage of golang side
+    milvus::proto::segcore::ClusteringCentroidsStats
+    CentroidsToPB(const T* centroids);
+
+    // transform flattened id mapping result to several PB files by each segment for future usage of golang side
+    std::vector<milvus::proto::segcore::ClusteringCentroidIdMappingStats>
+    CentroidIdMappingToPB(
+        const uint32_t* centroid_id_mapping,
+        const std::unordered_map<int64_t, int64_t>& segment_id_offset_map,
+        const std::map<int64_t, int64_t>& num_row_map);
+
+    std::shared_ptr<storage::MemFileManagerImpl> file_manager_;
```

Comment:
We could avoid adding more clustering-related utils to `FileManager`, but I think we should use filemanager to reuse the pull raw data function, so that we could still use one implementation for this util and won't be missed out in the future refactor.